File: python/tests/test_starter.py
import unittest

from starter import add, greet, project_name


class StarterTest(unittest.TestCase):
    def test_add(self) -> None:
        self.assertEqual(add(2, 3), 5)
        self.assertEqual(add(-1, 1), 0)

    def test_project_name(self) -> None:
        self.assertEqual(project_name(), "starter")

    def test_greet(self) -> None:
        self.assertEqual(greet("Ada"), "Hello, Ada")
        self.assertEqual(greet("  Ada  "), "Hello, Ada")
        self.assertEqual(greet(""), "Hello, world")
        self.assertEqual(greet("   "), "Hello, world")
        self.assertEqual(greet(None), "Hello, world")


if __name__ == "__main__":
    unittest.main()

File: c/tests/test_starter.c
#include "starter.h"

#include <stdio.h>
#include <string.h>

static int failures = 0;

static void expect(int ok, const char *message) {
    if (!ok) {
        fprintf(stderr, "FAIL %s\n", message);
        failures++;
    }
}

int main(void) {
    char buf[64];

    expect(starter_add(2, 3) == 5, "add");
    expect(starter_add(-1, 1) == 0, "add neg");
    expect(strcmp(starter_project_name(), "starter") == 0, "name");

    expect(starter_greet("Ada", buf, sizeof buf) == 0, "greet rc");
    expect(strcmp(buf, "Hello, Ada") == 0, "greet");
    expect(starter_greet("  Ada  ", buf, sizeof buf) == 0, "trim rc");
    expect(strcmp(buf, "Hello, Ada") == 0, "trim");
    expect(starter_greet("   ", buf, sizeof buf) == 0, "blank rc");
    expect(strcmp(buf, "Hello, world") == 0, "blank");
    expect(starter_greet(NULL, buf, sizeof buf) == 0, "null rc");
    expect(strcmp(buf, "Hello, world") == 0, "null");
    expect(starter_greet("Ada", buf, 4) == -1, "small");
    expect(starter_greet("Ada", NULL, 16) == -1, "null buf");

    if (failures != 0) {
        return 1;
    }

    puts("c ok");
    return 0;
}
